p2p/protocol/identify/id.go

```diff
@@ -156,19 +159,35 @@ func (ids *IDService) pushHandler(s inet.Stream) {
 }
 
 func (ids *IDService) Push() {
+	var wg sync.WaitGroup
+
+	// we could make this context timeout-less since we are only opening a new
+	// stream over an existing connection. This would avoid the need for the
+	// supervisory goroutine below, but timeout-less contexts in network operations
+	// make me nervous.
+	ctx, cancel := context.WithTimeout(ids.ctx, 15*time.Second)
+	ctx = inet.WithNoDial(ctx, "identify push")
+
 	for _, p := range ids.Host.Network().Peers() {
+		wg.Add(1)
 		go func(p peer.ID) {
-			ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
-			defer cancel()
+			defer wg.Done()
+
 			s, err := ids.Host.NewStream(ctx, p, IDPush)
 			if err != nil {
-				log.Debugf("error opening push stream: %s", err.Error())
+				log.Debugf("error opening push stream to %s: %s", p, err.Error())
 				return
 			}
 
 			ids.requestHandler(s)
```

Comment:
TODO: Not a bug in this PR but we should probably:

1. Do this in a goroutine.
2. Reset the stream if the context expires.

At the moment, we can hang the _entire_ push if a single peer stalls.

Reply:
Actually, this is now an issue because we're blocking everything else (which we weren't doing before).

Reply:
I can move the context back into the goroutine and remove the wait group, so that we don't stall.

Reply:
Note that the only reason I did this dance with the wait group was my desire to reuse the context instead of creating one for each peer.

Reply:
Ah, you want the `ids.requestHandler` call in a goroutine, so that we don't stall; much better than exploding the contexts.

Reply:
so resetting the stream will need an extra goroutine to look after it.

Reply:
It's probably more reasonable to wait in the dispatch goroutine for up to the context and reset the stream there.

Reply:
Also note we are not blocking anything, there is a supervisory goroutine waiting for the WaitGroup.

Reply:
We might simply get away with setting a `Deadline` in the stream.

Reply:
Done in https://github.com/libp2p/go-libp2p/pull/597/commits/5503cd7ff3cb79f9c1f2d3b11822a962352f79d4
I moved the `ids.requestHandler` into a goroutine and added a wait for the context cancellation to reset the stream if the handler doesn't complete by then.

Reply:
> Also note we are not blocking anything, there is a supervisory goroutine waiting for the WaitGroup.

Ah, got it. But better to reset the stream anyways.

Reply:
Yeah, let's not build up goroutines if we have a stuck peer.